torch/csrc/autograd/engine.cpp

```diff
@@ -742,44 +752,134 @@ auto Engine::execute(const edge_list& roots,
+  ready_queue(graph_task, at::kCPU).push(
       std::move(task), /* incrementOutstandingTasks */ false);
+  std::unique_lock<std::mutex> lock(print_mutex_);
+  LOG(INFO)<<"enqueued blocked task on CPU ready queue for graphTask: "<< graph_task;
+}
+
+std::shared_ptr<GraphTask> Engine::execute_node_task(const std::unique_ptr<NodeTask>& task) {
+  // local_graph_task represents the graph_task we retrieve from the queue.
+  // The outer graph_task represents the overall graph_task we need to execute
+  // for reentrant execution.
+  std::shared_ptr<GraphTask> local_graph_task;
+  if (!(local_graph_task = task->base_.lock())) {
+    // Reentrant thread's graph task should not expire since we hold a
+    // reference to it in this method.
+    LOG(INFO) << "GraphTask for function " << task->fn_->name()
+              << " is no longer valid, skipping execution";
+    return local_graph_task;
+  }
+
+  if (task->fn_ && !local_graph_task->has_error_.load()) {
+    AutoGradMode grad_mode(local_graph_task->grad_mode_);
+    try {
+      evaluate_function(local_graph_task, task->fn_.get(), task->inputs_);
+    } catch (std::exception& e) {
+      thread_on_exception(local_graph_task, task->fn_, e);
+    }
+  }
+
+  // Decrement the outstanding tasks.
+  --local_graph_task->outstanding_tasks_;
+
+  return local_graph_task;
+}
+
+void Engine::execute_until_ready_queue_empty(const std::shared_ptr<GraphTask>& graph_task) {
+  std::unique_lock<std::mutex> lock(print_mutex_);
+  std::shared_ptr<ReadyQueue> graph_task_rq = graph_task->cpu_ready_queue_;
+  // LOG(ERROR) << "Engine::execute_until_ready_queue_empty on graphTask: " << graph_task;
+  while(!graph_task_rq->empty()) {
+    auto task = graph_task_rq->pop();
+    execute_node_task(task);
+    LOG(INFO)<<"executed task fn: " << task->fn_->name()<< " instance: "<< task->fn_
+      << ", for graph_task: " << graph_task << " with outstanding task num:" << graph_task->outstanding_tasks_.load();
+  }
+  // Check if we've completed execution.
+  if (graph_task_completed(graph_task)) {
+    // We don't need to explicitly notify the owner thread, since
+    // 'mark_graph_task_completed' would mark the Future as completed and this
+    // would notify the owner thread that the task has been completed.
+    mark_graph_task_completed(graph_task);
+    LOG(INFO) <<"finished graph_task: " << graph_task;
+  } else {
+    // schedule a continuation
+    at::launch([this, graph_task]() {
+        execute_until_ready_queue_empty(graph_task);
+    });
+    LOG(INFO) << "scheduled a continuation on graph_task: "  << graph_task
+      << " with outstanding task num:" << graph_task->outstanding_tasks_.load();
+  }
+
 }
 
 std::shared_ptr<FutureVariableList> Engine::execute_with_graph_task(
     const std::shared_ptr<GraphTask>& graph_task,
-    std::shared_ptr<Node> graph_root) {
-  initialize_threads_pool();
+    std::shared_ptr<Node> graph_root,
+    bool async_mode) {
```

Comment:
With the current implementation, the "async_mode" argument is ignored when we are doing a reentrant call. Maybe this should actually be an error stating that this is not supported if we don't actually support it?